### PR TITLE
Added get_velocity function to EulerDiscreteScheduler.

### LINKED ISSUE #6359: EulerDiscreteScheduler not have get_velocity function and add_noise get error for prediction_type == "v_prediction"?
### Describe the bug

Finetune stable video diffusion but found two problems:

1. EulerDiscreteScheduler not have get_velocity function ?
2. EulerDiscreteScheduler  add_noise get error for prediction_type == "v_prediction"?

### Reproduction

1. EulerDiscreteScheduler not have get_velocity function ?
https://github.com/huggingface/diffusers/blob/main/examples/research_projects/controlnet/train_controlnet_webdataset.py#L1376

2. EulerDiscreteScheduler  add_noise get error for prediction_type == "v_prediction"?
https://github.com/huggingface/diffusers/blob/645a62bf3bcba47f97bf2a36a8689f40bd6b04fd/examples/research_projects/controlnet/train_controlnet_webdataset.py#L1341

### Logs

_No response_

### System Info

latest master branch

### Who can help?

@yiyixuxu @patrickvonplaten

### PR DESCRIPTION
**Pull request by RuiningLi**
# What does this PR do?

This PR adds the get_velocity function to EulerDiscreteScheduler, which is needed for training/fine-tuning with this scheduler if the `prediction_type` is `v_prediction`.

Fixes #6359 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@yiyixuxu

### CODE REVIEW

**Comment by yiyixuxu** [comment]
@RuiningLi 
I don't understand how this is related to https://github.com/huggingface/diffusers/pull/6395 ?

>Fixes https://github.com/huggingface/diffusers/pull/6395

**Comment by RuiningLi** [comment]
> @RuiningLi I don't understand how this is related to #6395 ?
> 
> > Fixes #6395

Sorry, I meant #6359 .

**Comment by yiyixuxu** [comment]
thanks!
I would like to understand the use case a little bit more 
Why do you need to use  `EulerDiscreteScheduler` for training?

**Comment by RuiningLi** [comment]
> thanks! I would like to understand the use case a little bit more Why do you need to use `EulerDiscreteScheduler` for training?

Stable Video Diffusion by default uses `EulerDiscreteScheduler`, so it would be nice to keep it consistent if we fine-tune the model.

**Comment by yiyixuxu** [comment]
@RuiningLi 
we normally use DDPM for training regardless of which scheduler it uses for inference by default 

I don't think we have a training script for SVD, do we?

**Comment by RuiningLi** [comment]
@yiyixuxu Not yet. Also, SVD itself uses `EulerDiscreteScheduler` for fine-tuning, and it also uses a continuous sigma sampler rather than a discrete one.

**Comment by bghira** [comment]
EulerDiscreteScheduler is used by **internal** training done by HuggingFace, for example, the SDXL ControlNet models were trained with `--use_euler`, but this means you can't train v-prediction models with that.. because the `scheduler.get_velocity()` is used if `prediction_type == "v_prediction"`

**Comment by RuiningLi** [comment]
> EulerDiscreteScheduler is used by **internal** training done by HuggingFace, for example, the SDXL ControlNet models were trained with `--use_euler`, but this means you can't train v-prediction models with that.. because the `scheduler.get_velocity()` is used if `prediction_type == "v_prediction"`

What do you mean exactly? If Huggingface internally use this scheduler to train, why does it prevent from having the `get_velocity` function?

**Comment by bghira** [comment]
that's not the point i was making, they use it internally on _epsilon_ prediction models. it's just that training on Euler isn't unordinary and it's even done inside HF itself. so supporting it in trainers doesn't seem unusual, either. it will unify the internal and example scripts.

**Comment by RuiningLi** [comment]
> that's not the point i was making, they use it internally on _epsilon_ prediction models. it's just that training on Euler isn't unordinary and it's even done inside HF itself. so supporting it in trainers doesn't seem unusual, either. it will unify the internal and example scripts. but this is required to merge before that can be done.

So do you mean that we need to first get this PR merged before we can have example training code for Euler v_prediction? If that is what you meant, then I agree with you!

**Comment by yiyixuxu** [comment]
I'm ok with adding this if there is a use case in training for it
@sayakpaul what do you think?

**Comment by sayakpaul** [comment]
Yes, indeed. This is an important thing to support as we move more and more to Karras-style training. IMO, we should support it.

**Review by sayakpaul — APPROVED** [review]
Thanks for your contributions!

**Comment by sayakpaul** [comment]
Good for me to merge once the test cases pass. For the quality tests, I think the error messages on the CI should be helpful.

File: src/diffusers/schedulers/scheduling_euler_discrete.py [hunk 1]
```diff
@@ -576,5 +576,44 @@ def add_noise(
         noisy_samples = original_samples + noise * sigma
         return noisy_samples
 
+    def get_velocity(
```
Comment (yiyixuxu) [inline src/diffusers/schedulers/scheduling_euler_discrete.py:579]: can we add a `#Copied from` statement?

Comment (yiyixuxu) [reply src/diffusers/schedulers/scheduling_euler_discrete.py:579]: see https://huggingface.co/docs/diffusers/v0.27.2/en/conceptual/contribution#copied-from-mechanism

Comment (RuiningLi) [reply src/diffusers/schedulers/scheduling_euler_discrete.py:580]: Added!

Comment (yiyixuxu) [reply src/diffusers/schedulers/scheduling_euler_discrete.py:580]: can you run `make fix-copies`?

Comment (RuiningLi) [reply src/diffusers/schedulers/scheduling_euler_discrete.py:580]: > can you run `make fix-copies`?

Can you help me run it? I don't have the installation available on the cluster.

Comment (sayakpaul) [reply src/diffusers/schedulers/scheduling_euler_discrete.py:579]: @yiyixuxu I took a look again. For EulerDiscrete the timesteps are not supposed long. So, we cannot use DDPM `get_velocity()` here I think.

Comment (RuiningLi) [reply src/diffusers/schedulers/scheduling_euler_discrete.py:579]: > @yiyixuxu I took a look again. For EulerDiscrete the timesteps are not supposed long. So, we cannot use DDPM `get_velocity()` here I think.

Sorry, my implementation is actually not copied from DDPM. In fact, it is not copied from anywhere. The implementation as is takes care of the continuous nature of the timestep.

**Review by yiyixuxu — APPROVED** [review]

File: src/diffusers/schedulers/scheduling_euler_discrete.py (hunk 1 above)
Comment (yiyixuxu) [reply src/diffusers/schedulers/scheduling_euler_discrete.py:579]: oh thanks!
yeah it looks good to me1
